{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"ARQUIVOS,ERROS E EXCEÇÕES (EXERCICIOS).ipynb","provenance":[],"collapsed_sections":[],"authorship_tag":"ABX9TyOq2J7ELDnmu2JD7kV+wJqa"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","metadata":{"id":"sieBlWs24uXd"},"source":["# **ARQUIVOS,ERROS E EXCEÇÕES**\n"]},{"cell_type":"markdown","metadata":{"id":"Uws-UWQ3DZOm"},"source":["**EXERCICIOS**"]},{"cell_type":"markdown","metadata":{"id":"sk0OpcapFilE"},"source":["1 - Faça um programa que leia um arquivo texto contendo uma lista de endereços IP e gere um outro arquivo, contendo um relatório dos endereços IP válidos e inválidos. O arquivo de entrada possui o seguinte formato: \n","\n","```\n","200.135.80.9\n","192.168.1.1\n","8.35.67.74\n","257.32.4.5\n","85.345.1.2\n","1.2.3.4\n","9.8.234.5\n","192.168.0.256\n","```\n","- O arquivo de saída possui o seguinte formato:\n","\n","```\n","[Endereços válidos:] \n","200.135.80.9 \n","192.168.1.1 \n","8.35.67.74 \n","1.2.3.4 \n","\n","[Endereços inválidos:] \n","257.32.4.5 \n","85.345.1.2 \n","9.8.234.5 \n","192.168.0.256 \n","\n","```\n"]},{"cell_type":"code","metadata":{"id":"UhVa7OIMFh11"},"source":["#codigo\n","import re\n","def regex(pattern, string):\n","\tcapt = re.match(pattern, string)\n","\tif bool(capt):\n","\t\treturn capt.group()\n","\treturn False\n","\n","ip_v = ''\n","ip_i = ''\n","list = {'0.25.33.19', '127.0.1.1','127.0.0.2','127.0.0.1','192.0.0.256','0.168.1.36','1.1.0.99','200.135.80.9','192.168.1.1','8.35.67.74','257.32.4.5','85.345.1.2','1.2.3.4','9.8.234.5','192.168.0.256'}\n","#podemos pegar essa lista dinamicamente se a caso necessario, como por exemplo puxar ela de um banco de dados \n","for ip in list:\n","\tif ip == '0.0.0.0':\n","\t\tip_v = ip_v + '\\n' + ip\n","\telif regex('^(([1-9]\\.|[1-9][0-9]\\.|1[0-9][0-9]\\.|2[0-4][0-9]\\.|25[0-5]\\.)([0-9]\\.|[1-9][0-9]\\.|1[0-9][0-9]\\.|2[0-4][0-9]\\.|25[0-5]\\.)([0-9]\\.|[1-9][0-9]\\.|1[0-9][0-9]\\.|2[0-4][0-9]\\.|25[0-5]\\.)([0-9]\\.|[1-9][0-9]\\.|1[0-9][0-9]\\.|2[0-4][0-9]\\.|25[0-5]\\.))$', ip + '.'):\n","\t\tip_v = ip_v + '\\n' + ip\n","\telse:\n","\t\tip_i = ip_i + '\\n' + ip\n","\n","print('[Endereços válidos:]')\n","print(ip_v)\n","print('[Endereços inválidos:]')\n","print(ip_i)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"V5V861cXHll3"},"source":["2 -Quando falamos sobre erros e exceções, algumas fontes de informação consideram que no processo de depuração, podemos encontrar três tipos de erro: erro de sintaxe, erro de tempo de execução e erro semântico. Defina o conceito de cada um deles \n"]},{"cell_type":"markdown","metadata":{"id":"qlNqaKUVBMDz"},"source":["A) Erro de sintaxe \n"]},{"cell_type":"code","metadata":{"id":"b7_3anxUHlFp"},"source":["#Primeiro irei colocar um exemplo de erro de sintaxe\n","\n","primt(\"ola\")  \n","\n","#Ou seja é um erro quando troca alguma letra de uma função ou variavel definida em seu codigo. (no caso aqui foi a letra m na função print)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"y4JkNGi-BUVx"},"source":["B) Erro de tempo de *execução*"]},{"cell_type":"code","metadata":{"id":"I37hLQl4BYkK"},"source":["#Erros de tempo de execução são produzidos se algo de errado acontece enquanto o programa está em execução. \n","#A maioria das mensagens de tempo de execução incluem informação sobre onde o erro ocorreu e que função estava em execução. Exemplo:\n","#Uma recursão infinita eventualmente causa um erro em tempo de execução identificado como \n"," maximum recursion depth exceeded \n","#(Excedida a máxima profundidade de recursão)."],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"s8mnTiKZBZ3J"},"source":["C) Erro semântico"]},{"cell_type":"code","metadata":{"id":"KNSFmzTJBg5p"},"source":["if (x==0) : # supondo que x chegue aqui com valor 1\n","  y = 2;\n","print(\"O valor e' nulo\");\n","#Entretanto, a mensagem é incorretamente impressa! \n","#A razão é que em Python a subordinação de comandos é definida por indentações, ou seja, \n","#para que o print(\"O valor e' nulo\") estivesse corretamente subordinado ao comando if ele precisaria estar alinhado ao comando y = 2;.\n","#Assim, correção é simples:\n","\n"," if (x == 0) :\n","   y = 2;\n","   print(\"O valor e' nulo\"); # correcao, alinhar com a linha acima"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"OCmNtxe1ITD8"},"source":["3 - Para explorar os erros e as exceções em Python, convido vocês a revisitarem os exercícios do primeiro módulo e para cada um deles, provocar alguns erros e avaliar as  mensagens do interpretador. Nesse estágio do nosso curso, deve estar mais claro e “menos aterrorizante” perceber que o Python nos ajuda a entender alguns dos erros que cometemos. Podemos implementar métodos de tratamento de exceções para as respectivas entradas de dados em cada um dos exercícios propostos. Você pode considerar a utilização dos seguintes recursos:\n","\n","\n"]},{"cell_type":"markdown","metadata":{"id":"ewLpfMVDSKEE"},"source":["A) Tente primeiramente utilizar os blocos try e except tratando as entradas de dados. Por exemplo: se o programa pede para que entre uma string, caso o usuário entre com um número inteiro, uma mensagem de erro deve aparecer e indicar o usuário o caminho correto."]},{"cell_type":"code","metadata":{"id":"VasZLCp9SJVN","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1620834125147,"user_tz":180,"elapsed":2673,"user":{"displayName":"3to","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GgL8VgI8p0kowu1avvJQlP64s6uRkis9cjJqIsvKg=s64","userId":"14877254573977447490"}},"outputId":"c27eb29f-1429-4e14-ec93-88ca3c0cb765"},"source":["try:\n","  a= int(input('Digite um numero: '))\n","except (ValueError, TypeError):\n","  print('Por favor insira apenas numeros inteiros.')"],"execution_count":17,"outputs":[{"output_type":"stream","text":["Digite uma letra: a\n"],"name":"stdout"}]},{"cell_type":"markdown","metadata":{"id":"RY8blAxOSSnz"},"source":["B) Tente tratar as exceções com mensagens especificas. Dessa forma, melhoramos a usabilidade do código."]},{"cell_type":"code","metadata":{"id":"sC0lzQmGSWr7"},"source":["try:\n","  a= int(input('Digite um numero: '))\n","except (ValueError, TypeError):\n","  print('Por favor insira apenas numeros inteiros.')\n","except KeyboardInterrupt:\n","  print('O usuario não informou os dados')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"My2TvKuNSXBz"},"source":["C) Tente utilizar quando possível, a construção completa envolvendo try, except, else e finally"]},{"cell_type":"code","metadata":{"id":"bf-d90ZwSca7"},"source":["try:\n","  a= int(input('Digite um numero: '))\n","  b= int(input('Digite outro numero: '))\n","  soma = a + b\n","except (ValueError, TypeError):\n","  print('Por favor insira apenas numeros inteiros.')\n","except KeyboardInterrupt:\n","  print('O usuario não informou os dados')\n","else:\n","  print(f'O resultado de é: {soma}')\n","finally:\n","  print('orbigado por usar nosso programa!')"],"execution_count":null,"outputs":[]}]}